### DISCUSSION TITLE
2 bimfile display

### DISCUSSION THREAD

**Discussion by paprophet** [iTwin/itwinjs-core#547] [Q&A] [answered]
now i usually convert revit files to one bim file by iTwin Snapshot,the rvt files have some reference file 
if a reference rvt file have change , i must convert all revit file to   one bim file by iTwin Snapshot  again
so i want just convert the  rvt file which have change and then display all bim file by imodeljs,
so How i can display 2 different bim file viewstate in one Viewport?

**Comment by shenshengyi** [comment]
async function addFeatureSymbology() {
  const vp = IModelApp.viewManager.selectedView;
  if (vp) {
    await addProvider(vp);
  }
}
async function removeFeatureSymbology() {
  const vp = IModelApp.viewManager.selectedView;
  if (vp) {
    await removeProvider(vp);
  }
}
/** 对TileTree的引用源自与用户打开的对象不同的IModelConnection */
class ExternalTreeRef extends TileTreeReference {
  private readonly _ref: TileTreeReference;
  private readonly _ovrs: FeatureSymbology.Overrides;

  public constructor(ref: TileTreeReference, ovrs: FeatureSymbology.Overrides) {
    super();
    this._ref = ref;
    this._ovrs = ovrs;
  }

  public get castsShadows() {
    return this._ref.castsShadows;
  }

  public get treeOwner() {
    return this._ref.treeOwner;
  }

  public addToScene(context: SceneContext): void {
    const tree = this.treeOwner.load();
    if (undefined === tree) return;

    // ###TODO transform
    const args = this.createDrawArgs(context);
    if (undefined === args) return;

    tree.draw(args);

    args.graphics.symbologyOverrides = this._ovrs;
    const branch = context.createBranch(args.graphics, args.location);
    context.outputGraphic(branch);
  }
}

class Provider implements TiledGraphicsProvider {
  private readonly _refs: TileTreeReference[] = [];
  public readonly iModel: IModelConnection;

  private constructor(
    vp: Viewport,
    iModel: IModelConnection,
    ovrs: FeatureSymbology.Overrides
  ) {
    this.iModel = iModel;
    for (const kvp of iModel.models.loaded) {
      const spatial = kvp[1].asSpatialModel;
      if (undefined !== spatial) {
        const ref = spatial.createTileTreeReference(vp.view);
        this._refs.push(new ExternalTreeRef(ref, ovrs));
      }
    }
  }

  public static async create(
    vp: Viewport,
    iModel: IModelConnection
  ): Promise<Provider> {
    const query = { from: SpatialModelState.classFullName, wantPrivate: false };
    const props = await iModel.models.queryProps(query);

    const modelIds = [];
    for (const prop of props) if (undefined !== prop.id) modelIds.push(prop.id);

    await iModel.models.load(modelIds);

    // Enable all categories (and subcategories thereof)
    const ecsql =
      "SELECT DISTINCT Category.Id as CategoryId from BisCore.GeometricElement3d WHERE Category.Id IN (SELECT ECInstanceId from BisCore.SpatialCategory)";
    const catIds: string[] = [];
    for await (const catId of iModel.query(ecsql))
      catIds.push(catId.categoryId);

    const subcatsRequest = iModel.subcategories.load(catIds);
    if (undefined !== subcatsRequest) await subcatsRequest.promise;

    // Ignore the symbology overrides defined on the viewport - instead, set up our own to draw our iModel's categories.
    const ovrs = new FeatureSymbology.Overrides();
    for (const catId of catIds) {
      const subcats = iModel.subcategories.getSubCategories(catId);
      if (undefined !== subcats)
        for (const subcat of subcats) ovrs.setVisibleSubCategory(subcat);
    }

    return new Provider(vp, iModel, ovrs);
  }

  public forEachTileTreeRef(
    _vp: Viewport,
    func: (ref: TileTreeReference) => void
  ): void {
    for (const ref of this._refs) func(ref);
  }
}
const providersByViewport = new Map<Viewport, Provider>();

/** A simple proof-of-concept for drawing tiles from a different IModelConnection into a Viewport. */
export async function addProvider(vp: Viewport): Promise<void> {
  const existing = providersByViewport.get(vp);
  if (undefined !== existing) {
    vp.dropTiledGraphicsProvider(existing);
    providersByViewport.delete(vp);
    await existing.iModel.close();
    return;
  }
  const filename = Config.App.getString("imjs_offline_imodel_Next");
  if (undefined === filename) return;
  let iModel;
  try {
    iModel = await SnapshotConnection.openFile(filename);
    const provider = await Provider.create(vp, iModel);
    providersByViewport.set(vp, provider);
    vp.addTiledGraphicsProvider(provider);
  } catch (err) {
    alert(err.toString());
  }
}
export async function removeProvider(vp: Viewport): Promise<void> {
  const existing = providersByViewport.get(vp);
  if (undefined !== existing) {
    vp.dropTiledGraphicsProvider(existing);
    providersByViewport.delete(vp);
    await existing.iModel.close();
  } else {
    alert("Provider不存在.");
  }
}

**Comment by shenshengyi** [comment]
> now i usually convert revit files to one bim file by iTwin Snapshot,the rvt files have some reference file
> if a reference rvt file have change , i must convert all revit file to one bim file by iTwin Snapshot again
> so i want just convert the rvt file which have change and then display all bim file by imodeljs,
> so How i can display 2 different bim file viewstate in one Viewport?

Hi, I realized the display of the contents of the two BIM files in the same viewport by customizing tiletreereference. The following is the relevant code, hoping to help you.

**Comment by pmconne** [comment, marked as answer]
@shenshengyi's code is copied from [display-test-app](https://github.com/imodeljs/imodeljs/blob/master/test-apps/display-test-app/src/frontend/TiledGraphics.ts). This is not a good solution for @paprophet's problem.

@paprophet A snapshot is a read-only representation of the state of an iModel at a specific point in time. If you are trying to keep an iModel in sync with .rvt files that are changing over time, you don't want a snapshot. Consider using [iTwin Synchronizer](https://www.bentley.com/en/products/product-line/digital-twins/itwin-synchronizer) instead of iTwin Snapshot.
